pep-9999.rst

```diff
@@ -0,0 +1,252 @@
+
+
+Proposal
+========
+
+The variance of the type parameters of a generic class can be specified
+with the ``+`` and ``-`` prefix operators on their respective type
+variables within the type parameter list of ``typing.Generic``, 
+``typing.Protocol``, or the generic base class. Instead of
+::
+
+    T_co = typing.TypeVar('T_co', covariant=True)
+    T_contra = typing.TypeVar('T_contra', contravariant=True)
+
+    class SupportsInvertOld(typing.Protocol[T_co]):
+        def __invert__(self) -> T_co: ...
+
+    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
+        def __le__(self, other: T_contra): ...
+
+
+The new syntax uses the ``+`` and ``-`` prefix operators to specify
+variadic type parameters in the same place where the generic type is
+declared:
+::
+
+    T = TypeVar("T")
+    
+    # covariant
+    class SupportsInvert(typing.Protocol[+T]):
+        def __invert__(self) -> T: ...
+
+    # contravariant
+    class SupportsPartialOrder(typing.Protocol[-T]):
+        def __le__(self, other: T): ...
+
+This syntax is inspired by Scala programming language [1]_.
+
+
+
+Specification
+=============
+
+
+Valid use locations
+-------------------
+
+
+The new type variable variance syntax can only be used within the type
+parameter list of ``typing.Generic``, ``typing.Protocol``, or a generic 
+base class. 
+
+The ``+`` and ``-`` prefix operators can be used on ``typing.TypeVar``
+and ``typing.ParamSpec`` instances, that are now already variadic, i.e.
+its ``__covariant__`` and ``__contravariant__`` attributes must be 
+``False``.
+
+When the same type variable is used on multiple generic base classes,
+they must share the same variance, e.g.
+::
+    
+    from typing import TypeVar, Callable, Container, Iterable, Protocol
+    
+    class LinkedList(Iterable[+T], Container[+T]): ...
+    class EventListener(Callable[[-T], None], Protocol[-T]): ...
+    
+are valid examples.
+
+
+Differences with current syntax
+-------------------------------
+
+The new typevar operators return a transparent wrapper around the 
+original type variable, which can be accessed with the ``__origin__``
+attribute on the returned wrapper. e.g.::
+
+    (+T).__origin__ is T
+    (+T).__covariant__ is True
+    (+T).__contravariant__ is False
+    (+T).__name__ == T.__name__
+    (+T).__constraints__ == T.__constraints__
+    (+T).__bound__ is T.__bound__
+
+
+Thus, type variables defined with ``covariant=True`` and 
+``contravariant=True``, are not equivalent to ``+T`` and ``-T``.
+
+
+``+T`` and ``-T`` are not valid type annotations, and should only be
+used within the generic type parameter list of generic base classes, e.g.:: 
+
+    class Spam(typing.Generic[+KT]): ...
+    class Eggs(typing.Protocol[-KT, +VT]): ...
+    class HamSet(typing.Sequence[+T]): ...
+
+are valid uses.
+
+All variance rules that apply to user-defined generic types should apply
+in the same way with the new syntax, as they do with the current syntax,
+and vice-versa.
+
+
```

Comment:
Should we outline a deprecation plan for variance parameters in `typing.TypeVar`?

Reply:
I don't think so. PEP 604 hasn't done so either.

Reply:
My views are likely biased, since I work on tooling consuming type annotations, but I would much prefer for both PEP-604 and [PEP-613](https://www.python.org/dev/peps/pep-0613/) to outline a deprecation plan for the feature they improve on. Otherwise, tooling authors are left in a situation where they have to support multiple variations of a feature forever.

Happy to bring this up on typing-sig@ independently from this PEP.